src/osemgrep/cli_scan/Core_runner.ml

```diff
@@ -101,13 +101,33 @@ let group_rules_by_target_language rules : (Xlang.t * Rule.t list) list =
                 Hashtbl.replace tbl lang (rule :: rules)));
   Hashtbl.fold (fun lang rules acc -> (lang, rules) :: acc) tbl []
 
+let get_targets_with_unknown_extension all_targets =
+  (* Note: we could have this function also split out the targets
+     with unknown extensions, but it's probably a fairly minor
+     optimization that would make the code more complicated *)
+  let explicit_targets =
+    all_targets
+    |> Common.map_filter (fun (target, info) ->
+           match info with
+           | Target_file.NoSpecialTargeting -> None
+           | Target_file.ExplicitFileScanUnknownExtension -> Some target)
+  in
+  explicit_targets
+  |> List.filter (fun x ->
+         not (Filter_target.filter_target_has_known_extension x))
+
 let split_jobs_by_language all_rules all_targets : Lang_job.t list =
+  let targets_with_unknown_extension =
+    get_targets_with_unknown_extension all_targets
+  in
+  let all_targets = Target_file.fpaths_of_target_files all_targets in
   all_rules |> group_rules_by_target_language
   |> Common.map_filter (fun (xlang, rules) ->
          let targets =
            all_targets
            |> List.filter (Filter_target.filter_target_for_xlang xlang)
          in
+         let targets = targets @ targets_with_unknown_extension in
          if Common.null targets then None
          else Some ({ xlang; targets; rules } : Lang_job.t))
```

Comment:
I would stay away from the language "has unknown extension" because it's overly restrictive. The model we have is:

* a list of scannable targets (result of global filtering involving semgrepignore and such)
* a list of rules
* a way to determine if a rule applies to a given target file: this is the part that should not assume that file extensions are necessary or sufficient to determine if a rule applies to a file. Note also that I'm avoiding the term "language" because this also has been another source of confusion.

Maybe we should deprecate `--scan-unknown-extensions` and rename it to something correct such as `--ignore-file-type` or something else that doesn't confuse the user with notions of extensions or languages that don't apply in the most general case.

Reply:
> Maybe we should deprecate --scan-unknown-extensions and rename it to something correct such as --ignore-file-type

Yes! I confused myself multiple times in the implementation of `--scan-unknown-extensions` because it was unclear intuitively what it did, and the docs are inaccurate. I don't think we should handle that during the port, but this flag is definitely confusing